### PR TITLE
fix(nervous_system): Fix a bug with advancing buf

### PR DESCRIPTION
**Pull request by jasonz-dfinity**
# Root Cause

When decoding canister state from protobuf during `post_upgrade`, a buffered reader is used for reducing the heap memory usage while decoding. In an edge case where `Buf::advance(0)` is called by prost library, we have a divide-by-zero situation.

# Trigger

This seems to only happen when prost skips some singleton field at the end (in terms of protobuf tag number). The situation is reproduced by adding a `bool b = 3;` protobuf field to `TestMessage`. This is similar to `SnsRootCanister` removing `updated_framework_canisters_memory_limit`.

# Fix

Calling `Buf::advance(0)` should always be a no-op, so an early return is performed. Also, the reasoning about the `checked_div_mod` succeeding is documented.

# Other Changes

* An `assert!` on the buffer size is added.
* The error message "Failed to calculate quotient" was not very helpful. The `checked_div_mod` also diverges from the pattern that `checked_` arithmetic does not panic itself, but returns `Option<T>` if arithmetic error happens. Now the panic happens at the call site.

### CODE REVIEW

**Review by max-dfinity — APPROVED** [review]
Please add a regression test if possible before merging, but changes look okay.

File: rs/nervous_system/common/src/memory_manager_upgrade_storage.rs
Comment (aterga) [inline rs/nervous_system/common/src/memory_manager_upgrade_storage.rs:329]: Seems like there's undesirable code duplication between here and rs/nervous_system/common/src/dfn_core_stable_mem_utils.rs

Is there a simple way to avoid that / a good reason to duplicate the code?

**Review by max-dfinity — APPROVED** [review]

File: rs/nervous_system/common/src/memory_manager_upgrade_storage.rs
Comment (jasonz-dfinity) [reply rs/nervous_system/common/src/memory_manager_upgrade_storage.rs:329]: True. Although, I think creating a common abstraction for both sides is difficult, mainly because they depend on slightly different StableMemory layers to read the memory. It might be possible to simply refactor some common logic to a few functions, but I imagine the reader would have to deal with 3 modules instead of 2, since there is still parts that cannot be shared but are non-trivial. I'm not convinced that would be better.

Comment (aterga) [reply rs/nervous_system/common/src/memory_manager_upgrade_storage.rs:329]: Okay, let's not block on this, then

**Comment by jasonz-dfinity** [comment]
> Please add a regression test if possible before merging, but changes look okay.

Discussed offline, the unit tests with `TestMessage*` already reproduces the issue (which fails without the fix)

**Pull request merged**
